### PR TITLE
Kernel: Fix IRQ problems

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

**Comment by supercomputer7** [comment]
> > Kernel: Allow IRQController class to be RefCounted
> 
> Hmm? Do we really want the `IRQController` to be destroyed when the last `IRQHandler` is destroyed?

AFAIK It won't since we have a RefPtr to it in InterruptManagement class, and we give it with `get_responsible_irq_controller()` method.

**Comment by awesomekling** [comment]
> > > Kernel: Allow IRQController class to be RefCounted
> > 
> > 
> > Hmm? Do we really want the `IRQController` to be destroyed when the last `IRQHandler` is destroyed?
> 
> AFAIK It won't since we have a RefPtr to it in InterruptManagement class, and we give it with `get_responsible_irq_controller()` method, but I could be wrong about this.

What I mean is.. the purpose of ref-counting is to keep an object alive as long as one of its owners is alive. So by making this ref-counted, we are saying that `IRQController` should be deleted if all `InterruptManagement` and `IRQHandler` objects are deleted. Does this ever happen? Does `IRQController` really need to be deletable? It's not like we want to switch back and forth between different controllers at runtime, right?

**Comment by supercomputer7** [comment]
> > > > Kernel: Allow IRQController class to be RefCounted
> > > 
> > > 
> > > Hmm? Do we really want the `IRQController` to be destroyed when the last `IRQHandler` is destroyed?
> > 
> > 
> > AFAIK It won't since we have a RefPtr to it in InterruptManagement class, and we give it with `get_responsible_irq_controller()` method, but I could be wrong about this.
> 
> What I mean is.. the purpose of ref-counting is to keep an object alive as long as one of its owners is alive. So by making this ref-counted, we are saying that `IRQController` should be deleted if all `InterruptManagement` and `IRQHandler` objects are deleted. Does this ever happen? Does `IRQController` really need to be deletable? It's not like we want to switch back and forth between different controllers at runtime, right?

Probably not a good idea to do a switch. However, when doing a shutdown/reboot, `IRQController` should mask all interrupts, so we can do that in the destructor.

**Comment by supercomputer7** [comment]
This PR currently have a workaround for #1321, but doesn't really fix it completely.
The E1000 network adapter creates Spurious IRQs under heavy loads. I'd guess it's related to the code in the E1000 driver, but I don't know for sure.

File: Kernel/Interrupts/IRQController.h [hunk 1]
```diff
@@ -36,7 +37,7 @@ enum class IRQControllerType {
     i82093AA = 2 /* Intel 82093AA I/O ADVANCED PROGRAMMABLE INTERRUPT CONTROLLER (IOAPIC) */
 };
 
-class IRQController {
+class IRQController : public RefCounted<IRQController> {
```
Comment (awesomekling) [inline Kernel/Interrupts/IRQController.h]: When we make something ref-counted, we must make sure to `adopt` the initial ref wherever we do `new` (or it is guaranteed to leak.)

```cpp
auto controller = adopt(*new IRQController);
```

**Pull request merged**